dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -14,19 +15,35 @@
 public class RelationAnalysis {
 
     private RelationAnalysis(VerificationTask task, Context context, Configuration config) {
+        context.requires(ExecutionAnalysis.class);
         context.requires(AliasAnalysis.class);
-        context.requires(BranchEquivalence.class);
+        context.requires(Dependency.class);
         context.requires(WmmAnalysis.class);
-        run(task, context);
     }
 
-    public static RelationAnalysis fromConfig(VerificationTask task, Context context, Configuration config) throws InvalidConfigurationException {
-        return new RelationAnalysis(task, context, config);
+    /**
+     * Performs a static analysis on the relationships that may occur in an execution.
+     * @param memoryModel Collection of relations and axioms to perform this analysis on.
+     * @param task Program, target memory model and property to check.
+     *             The target memory model does not have to coincide with {@code memoryModel}.
+     * @param context Collection of static analyses already performed on {@code task} with respect to {@code memoryModel}.
+     *                Should at least include the following elements:
+     *                <ul>
+     *                    <li>{@link ExecutionAnalysis}
+     *                    <li>{@link Dependency}
+     *                    <li>{@link AliasAnalysis}
+     *                    <li>{@link WmmAnalysis}
+     *                </ul>
+     * @param config User-defined options to further specify the behavior.
+     */
+    public static RelationAnalysis fromConfig(Wmm memoryModel, VerificationTask task, Context context, Configuration config) throws InvalidConfigurationException {
+        RelationAnalysis a = new RelationAnalysis(task, context, config);
+        a.run(memoryModel, task, context);
+        return a;
     }
 
-    private void run(VerificationTask task, Context context) {
+    private void run(Wmm memoryModel, VerificationTask task, Context context) {
         // Init data context so that each relation is able to compute its may/must sets.
-        Wmm memoryModel = task.getMemoryModel();
         for (Axiom ax : memoryModel.getAxioms()) {
             ax.getRelation().updateRecursiveGroupId(ax.getRelation().getRecursiveGroupId());
         }
```

Comment:
I think this code is dangerous, cause there is a mismatch between the provided memory model and the memory model inside the task.
That being said, I don't think any relation ever accesses `task.getMemoryModel` so they won't see any mismatch.
I understand that you did this to handle `Refinement`, but now that `VerificationTask` is a simple data container, I think it is reasonable to have the `RefinementSolver` construct a new `VerificationTask` object with the correct memory model and pass that one to the relation analysis (we can extend the `Builder` to allow creating copies of a preexisting task with minor changes).
I also think that this makes intuitive sense: The task that gets encoded by `RefinementSolver` is fundamentally a different one to the one that it will eventually solve.